src/transport/CertificatePairingSession.cpp

```diff
@@ -0,0 +1,409 @@
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR CertificatePairingSession::Pair(const Transport::PeerAddress peerAddress, NodeId myNodeId, NodeId peerNodeId,
+                                           uint16_t myKeyId, PairingSessionDelegate * delegate)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    SuccessOrExit(err);
+
+    mConnectionState.SetPeerAddress(peerAddress);
+    mConnectionState.SetPeerNodeId(peerNodeId);
+
+    err = SendSigmaR1();
+    SuccessOrExit(err);
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        Clear();
+    }
+    return err;
+}
+
+CHIP_ERROR CertificatePairingSession::DeriveSecureSession(const uint8_t * info, size_t info_len, SecureSession & session)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    VerifyOrExit(info != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(info_len > 0, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(mPairingComplete, err = CHIP_ERROR_INCORRECT_STATE);
+
+    // ToDo: Use asymmetric keypair to create the secure session
+    // err = session.Init();
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR CertificatePairingSession::SendSigmaR1()
+{
+    uint16_t data_len = 0;
+
+    System::PacketBufferHandle msg_R1;
+
+    msg_R1 = System::PacketBuffer::NewWithAvailableSize(data_len);
+    VerifyOrReturnError(!msg_R1.IsNull(), CHIP_SYSTEM_ERROR_NO_MEMORY);
+
+    // ToDo: Construct SigmaR1 message in form of the following structure
+    /*
+    struct SigmaR1
+    {
+        uint8_t random[32];
+        uint8_t sessionid[2];
+        TrustedRootIdentifier trusted_roots[];
+        CipherSuite cipher_suites[];
+        KeyShare key_shares[];
+        Extensions extensions[];
+    };
+    */
+
+    msg_R1->SetDataLength(data_len);
+    mNextExpectedMsg = Protocols::SecureChannel::MsgType::CASE_SigmaR2;
+
+    // Call delegate to send the msg to peer
+    ReturnErrorOnFailure(AttachHeaderAndSend(Protocols::SecureChannel::MsgType::CASE_SigmaR1, std::move(msg_R1)));
+
+    ChipLogDetail(Inet, "Sent SigmaR1 msg");
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CertificatePairingSession::HandleSigmaR1_and_SendSigmaR2(const PacketHeader & header,
+                                                                    const System::PacketBufferHandle & msg)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    uint16_t data_len = 0;
+
+    const uint8_t * buf = msg->Start();
+
+    System::PacketBufferHandle msg_R2;
+
+    ChipLogDetail(Inet, "Received SigmaR1 msg");
+
+    VerifyOrExit(buf != nullptr, err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+
+    // ToDo: Verify received SigmaR1 message.
+
+    mConnectionState.SetPeerKeyID(header.GetEncryptionKeyID());
+
+    msg_R2 = System::PacketBuffer::NewWithAvailableSize(data_len);
+    VerifyOrExit(!msg_R2.IsNull(), err = CHIP_SYSTEM_ERROR_NO_MEMORY);
+
+    // ToDo: Construct SigmaR2 message in form of the following structure
+    /*
+    struct SigmaR2Encrypted
+    {
+        opaque responder_identity;
```

Comment:
Note that spec team is actively working on expressing the serialization of all of these fields as a wire format *as we speak*.

@balducci-apple @bzbarsky-apple 

A notable aspect is that CASE is expected to use TLV-encoded encoded certificates (@emargolis can point to the related bits) for operational certs.